cmd/utils/jeth.go

```diff
@@ -129,76 +110,62 @@ func (self *Jeth) NewAccount(call otto.FunctionCall) (response otto.Value) {
 }
 
 func (self *Jeth) Send(call otto.FunctionCall) (response otto.Value) {
-	reqif, err := call.Argument(0).Export()
-	if err != nil {
-		return self.err(call, -32700, err.Error(), nil)
+	ro := call.Argument(0).Object()
+	if ro == nil || (ro.Class() != "Array" && ro.Class() != "Object") {
+		throwJSExeception("Internal Error: request must be an object or array")
 	}
-
-	jsonreq, err := json.Marshal(reqif)
-	var reqs []rpc.JSONRequest
-	batch := true
-	err = json.Unmarshal(jsonreq, &reqs)
+	reqv, err := call.Otto.Call("JSON.stringify", otto.NullValue(), ro)
 	if err != nil {
-		reqs = make([]rpc.JSONRequest, 1)
-		err = json.Unmarshal(jsonreq, &reqs[0])
-		batch = false
+		throwJSExeception(err.Error())
 	}
-
-	call.Otto.Set("response_len", len(reqs))
-	call.Otto.Run("var ret_response = new Array(response_len);")
-
-	for i, req := range reqs {
-		err := self.client.Send(&req)
-		if err != nil {
-			return self.err(call, -32603, err.Error(), req.Id)
-		}
-
-		result := make(map[string]interface{})
-		err = self.client.Recv(&result)
-		if err != nil {
-			return self.err(call, -32603, err.Error(), req.Id)
-		}
-
-		_, isSuccessResponse := result["result"]
-		_, isErrorResponse := result["error"]
-		if !isSuccessResponse && !isErrorResponse {
-			return self.err(call, -32603, fmt.Sprintf("Invalid response"), new(int64))
-		}
-
-		id, _ := result["id"]
-		call.Otto.Set("ret_id", id)
-
-		jsonver, _ := result["jsonrpc"]
-		call.Otto.Set("ret_jsonrpc", jsonver)
-
-		var payload []byte
-		if isSuccessResponse {
-			payload, _ = json.Marshal(result["result"])
-		} else if isErrorResponse {
-			payload, _ = json.Marshal(result["error"])
-		}
-		call.Otto.Set("ret_result", string(payload))
-		call.Otto.Set("response_idx", i)
-
-		response, err = call.Otto.Run(`
-		ret_response[response_idx] = { jsonrpc: ret_jsonrpc, id: ret_id, result: JSON.parse(ret_result) };
-		`)
+	req := json.RawMessage(reqv.String())
+
+	if cb := call.Argument(1); cb.IsFunction() {
+		// Asynchronous call: do the the communication on another goroutine, then switch
+		// back to the event loop and call cb with the response or error.
+		go self.execRequests(&req, func(resp json.RawMessage, err error) {
```

Comment:
If you work on a solution keep in mind that we also have subscriptions. I'm not sure how these will be implemented in web3 therefore I didn't change anything yet.